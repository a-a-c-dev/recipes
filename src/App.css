*{
    margin: 0;
    padding: 0;
    box-sizing: border-box;
}
@font-face {
    font-family: primaryFont;
    src: url(./font/JosefinSlab-VariableFont_wght.ttf);
  }

html{
    font-size: 62.5%
}

body{
    font-size: 1.4rem;

}





.app{
    min-height: 100vh;
    margin: auto;    
    position: relative;
    overflow: hidden;
    padding: 1vh 4vw 2vh;
    background-color: rgb(66 66 66 / 40%);
    font-family: "primaryFont";

}
.video-bg{ 
    z-index: -1;
    overflow: hidden;
    background-color: rgba(0,0,0,.2);
    width: 100vw;
    height: 100vh;
    object-fit: cover;
    position: fixed;
    top: 0;
    left: 0;
    bottom: 0;
  }
.app-container{
      max-width: 1440px;
      margin: 0 auto;
}
.app-header-container{
    font-family: "primaryFont";
    padding: 10em 0 0;
    margin: 0;
    color:linear-gradient(to top, #bb5c5c9e 0% ,#fffcfc 40%, #f5d100 100%);
    
    display:flex;
    flex-direction: column;
    align-items:center; 
    justify-content:center; 
    text-align: center;
}
.app-header-container h1{
    font-size: 5.5rem;
    line-height: 1.2;
    letter-spacing: .75rem;
    font-weight: 900;
    padding: .5em 0 ;
    text-align: center;
    background-image:linear-gradient(to top, rgb(250, 250, 137) 0% ,#f8f4ce 40%, rgb(255, 255, 231) 70%, #f5d100 100%);
    -webkit-background-clip: text;
    -webkit-text-fill-color: transparent;
    text-transform: uppercase;
}
.app-header-container h3{
    font-weight: 700;
    font-size: 3.85rem;
    line-height: 1.4;
    background-color: rgba(0,0,0,.02);    
    background-image: linear-gradient(to top, rgb(250, 250, 137) 0% ,#f8f4ce 40%, rgba(187,135,118,1) 70%, #f5d100 100%);
    -webkit-background-clip: text;
    -webkit-text-fill-color: transparent;
    border: none;
}

.search-form{
    min-height: 10vh;
    display: flex;
    align-items: center;
    justify-content: center;
    border-radius: 12px;
    max-width: 350px;
    margin:3em auto;
    position: relative;
}
.validation-error{
    width: 100%;
    font-weight: 700;
    position: absolute;
    top:100%;
    color: #ffffff;
    font-size: 2.5rem;
    text-align: center;
    background-color: #ff0000;
    border-radius:5px ;

}
.search-bar{
    width: 100%;
    border: none;
    padding: 1em ;
    letter-spacing: 1.47px;
    font-size: 1.85rem;
    border-top-left-radius:5px;
    border-bottom-left-radius:5px;
}
.search-button{
    background: rgb(135, 151, 156);
    border:none;
    padding:1em;
    font-size: 1.85rem;
    letter-spacing: 0.15rem;
    color:white;
    border-top-right-radius: 5px;
    border-bottom-right-radius: 5px;
}
.recipes{
    display: flex;
    justify-content: space-around;
    flex-wrap: wrap;

}

.error{
    display: flex;
    flex-wrap: wrap;
    align-items: center;
    justify-content: center;
    text-align: center;
    width: 100%;
    max-width: 300px;
    text-transform: capitalize;
    margin: 0 auto;
    background: rgba(66, 66, 66, 0.727);
}
.icon-container{
    max-width: 50px;
}

.icon-container img{
    width: 100%;
    object-fit: contain;
}
.error h3{
    margin: 0 0 0 .5rem ;
    font-size: 32px;
    line-height: 1.4;
    font-family: "primaryFont";
    color: #b10000;
} 


@media (max-width: 768px) {
    html{
        font-size: 55%
    }
    
 }

 @media (min-width:769px) and (max-width: 1024px) {
    .app-header-container{
        padding: 2em;
    }
    html{
        font-size: 57.5%
    }
 }