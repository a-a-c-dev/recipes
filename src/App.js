import React, { useEffect, useState,Suspense,lazy, useMemo } from 'react';
import { Header } from './components/Header';
import {BgVideo} from  './components/BgVideo';
import {Error}  from './components/Error';
import Search from './components/Search';
import {Spinner} from './components/Spinner';
import './App.css';
import LocalStorageManager from "./utils/LocalStorageManager";

const Recipes = lazy(() => import ('./components/Recipes'));

const App = () => {
  const APP_ID = process.env.REACT_APP_RECIPES_API_ID;
  const APP_KEY = process.env.REACT_APP_RECIPES_API_KEY;
  const storageQuery = LocalStorageManager.get('searchQuery');
  const storageRecipes = (!LocalStorageManager.get('recipes')?[]:LocalStorageManager.get('recipes'));
  const [recipes, setRecipes] = useState(storageRecipes);
  const [query, setQuery] = useState(storageQuery);
  const [isLoading,setIsLoading] = useState(true);
  const[err,setError] =useState('');
  const[hasError,setHasError] =useState(false);
  let url = `https://api.edamam.com/search?q=${query}&app_id=${APP_ID}&app_key=${APP_KEY}`;

 

  const addRecipes = (e,searching,isValid) => {
    e.preventDefault();
    if(!isValid)return
    setQuery(searching);
  }

  const getRecipes = useMemo(()=>
    async () => {
    try{  
      const response = await fetch(url);
      const data = await response.json();
      if(!data.hits.length>0)setError("couldnt found the recipe, please search for another")
      setRecipes(data.hits);
      LocalStorageManager.set("recipes",data.hits)
    }
    catch(error){
      setHasError(true);
    }
},[url]);
  
  useEffect(() => {
    if (!recipes.length>0) setError("Still loading, please wait")
    if (hasError) setError("Something happend, please reload the page and check you Internet connection")
  }, [recipes.length,isLoading, hasError]);
  
  
  useEffect(() => {
    getRecipes();
    setIsLoading(true);
  }, [query,getRecipes]);
  
  
  useEffect(() => {
    if(!recipes.length>0 && !isLoading)setError("Couldn`t find the recipes, please serach again");
    if(recipes.length>0)setIsLoading(false);
  }, [ recipes, isLoading]);



  return (
    <div className="app">
      <BgVideo/>
      <div className="app-container">
        <Header/>
        <Search addRecipes={addRecipes}/>
        <Suspense fallback={<Spinner/>}>
        <Recipes  recipes={recipes} />
        </Suspense>
      </div> 
      {isLoading? 
        <Error err={err}/>
        :null
      }
    </div>
  );
}

export default App;
