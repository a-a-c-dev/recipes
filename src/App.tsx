import React, { useEffect, useState,Suspense,lazy, useMemo } from 'react';
import { Header } from './components/Header';
import {BgVideo} from  './components/BgVideo';
import {Error}  from './components/Error';
import Search from './components/Search';
import {Spinner} from './components/Spinner';
import './App.css';
import LocalStorageManager from "./utils/LocalStorageManager";
import useLocalStorage from './hooks/useLocalStorage';

const Recipes = lazy(() => import ('./components/Recipes'));
interface RecipeData {
  hits: {
    recipe: {
      calories: number;
      label: string;
      image: string;
      url: string;
      ingredients: string[];
      uri: string;
    };
  }[];
}
const App = () => {
  const APP_ID = process.env.REACT_APP_RECIPES_API_ID;
  const APP_KEY = process.env.REACT_APP_RECIPES_API_KEY;
  const [recipes, setRecipes] = useLocalStorage('recipes',!LocalStorageManager.get('recipes')?[]:LocalStorageManager.get('recipes'));
  const [query, setQuery] = useLocalStorage('searchQuery', LocalStorageManager.get('searchQuery')|| 'pasta');
  const [isLoading,setIsLoading] = useState(true);
  const[err,setError] =useState('');
  const[hasError,setHasError] =useState(false);
  let url = `https://api.edamam.com/search?q=${query}&app_id=${APP_ID}&app_key=${APP_KEY}`;

 

  const addRecipes = (e:React.FormEvent<HTMLFormElement>,searching:string,isValid:boolean) => {
    e.preventDefault();
    if(!isValid)return
    setQuery(searching);
  }

  const getRecipes = useMemo(()=>async () => {
    setIsLoading(true);
    try{  
      const response = await fetch(url);
      const data:RecipeData = await response.json();
      if(data.hits.length>0){
     //   let newRecipes: { recipe: { calories: number, label: string, image: string, url: string, ingredients: string[], uri: string } }[] = data.hits.map(({ recipe: { calories, label, image, url, ingredients, uri } }) => ({ recipe: { calories, label, image, url, ingredients, uri } }));
     const newRecipes = data.hits.map(({ recipe}) => ({
          recipe: {
            calories: recipe.calories,
            label: recipe.label,
            image: recipe.image,
            url: recipe.url,
            ingredients: recipe.ingredients,
            uri: recipe.uri,
          },
        }));
        setRecipes(newRecipes);       
        setIsLoading(false);
      }
      else setError("couldnt found the recipe, please search for another")
    }
    catch(error){
      setHasError(true);
    }
},[url,setRecipes]);
  
  useEffect(() => {
    if (!(recipes.length>0)) setError("Still loading, please wait")
    if (hasError) setError("Something happend, please reload the page and check you Internet connection" )
    if(recipes.length>0)setIsLoading(false);
    return () => { 
      setError("")
      setIsLoading(false)
    }
  }, [recipes.length,isLoading, hasError]);
  
  
  useEffect(() => {
    let isMounted = true;
    if(isMounted)getRecipes(); 
    return () => { isMounted = false}
  }, [query,getRecipes]);

  return (
    <div className="app" >
      <BgVideo/>
      <div className="app-container">
        <Header/>
        <Search addRecipes={addRecipes}/>
        {recipes && <Suspense fallback={<Spinner/>}>
            <Recipes  recipes={recipes} />
          </Suspense>
        }
        {err.length?<Error err={err}/>:null}
      </div> 
    </div>
  );
}

export default App;
