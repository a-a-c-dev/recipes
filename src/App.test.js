import React from 'react';
import { mount, shallow } from 'enzyme';
import App from './App';

describe('Testing app component', () => {
    let wrapper;
    let OnChangeSearching = jest.fn();
    beforeEach(() => {
        wrapper = shallow(<App onChange={OnChangeSearching} />);
    });


    it('Should render correctly ', () => {
        expect(wrapper).toMatchSnapshot();
    });

    xit('simulate change handler', () => {
        //wrapper.find('input').instance().onChange({ target: { value } });
        wrapper.find('input').simulate('change', { target: { value: 'test' } });
        // expect(OnChangeSearching).toBeCalledWith('test');
        //expect(mockFn.mock.calls[0][0]).toBe('test');

    });
    afterEach(() => {
        wrapper.unmount();
    });

});


describe('', () => {
    let wrapper;
    let OnChangeSearching = jest.fn();
    const props = {
        onChange: OnChangeSearching
    }
    beforeEach(() => {
        wrapper = shallow(<App {...props} />);
    });

    it('Should test change handler', () => {
        let value = '';
        const mockCallback = OnChangeSearching.mock.calls.length;
        expect(mockCallback).toBe(0);
        expect(wrapper.find('input').at(0).prop('value')).toEqual('');
        wrapper.find('input').props().onChange({ target: { value: 'test' } });
        wrapper.find('input').simulate('change', { target: { value: 'test' } });
        wrapper.find('input').at(0).props().onChange({ e.target:{ e.value: 'test'}}})
    .onChange({ City: { target: { value: 'test' } } })
//expect(mockCallback).toBe(1);
expect(wrapper.find('input').at(0).prop('value')).toEqual('');

expect(wrapper.find('input').at(0).prop('value')).toEqual('test');
        //expect(wrapper.find({ prop: 'value' })).to.have.lengthOf(1)

        //expect(OnChangeSearching).toHaveBeenCalled();
        //expect(OnChangeSearching).toHaveBeenCalledWith("test");



    })

afterEach(() => {
    wrapper.unmount();
});
})