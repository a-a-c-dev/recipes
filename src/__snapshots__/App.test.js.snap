// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing app component Should render correctly  1`] = `ShallowWrapper {}`;
